brave/brave-core
# private search engines settings option

fix https://github.com/brave/brave-browser/issues/23091

Search engines setting has separate option for configuring
search provider for private profile.

`PrivateWindowSearchEngineProviderService` configures search provider when user change settings.
`GuestWindowSearchEngineProviderService` is deleted as it's already dead code.
With this PR, DDG toggle in Private NTP will not be visible anymore. Instead, user can configure its provider via settings.


Resolves 

![image](https://user-images.githubusercontent.com/6786187/172594406-ce9957a5-1168-4fe3-b3d2-18f0640e89fa.png)

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
`npm run test brave_browser_tests --filter=SearchEngineProviderServiceTest*`
`npm run test brave_browser_tests --filter=ExtensionBrowserTest.ExtensionSearchProviderWithPrivateWindow`

**1. Test with clean profile or existing user that doesn't turn DDG toggle on in private NTP**
1. Launch Brave
2. Go to brave://settings/search and check same provider is set as Normal Window & Private Window's initial provider
3. Check Each option has same provider list
4. Change private window's provider and check it's used as a default provider in private window
5. Check private window's option change doesn't affect normal or private tor window
6. Go to `Manage search engines and site search` sub section and modify any provider's property (ex, name)
7. Check provider list has also updated list about the changes done at step 6
8. Go to brave://settigns/reset and do reset. Check private provider option is also has default one

**2. Test with existing user that turn DDG toggle on in private NTP**
1. Prepare profile that DDG toggle in Private NTP is on (See below how to prepare this profile)
2. Launch Brave
3. Go to brave://settings/search and check DDG is set as private window's provideer
4. Check DDG toggle button is not visible anymore in private NTP
5. User will control search provider for private profile via settings


**3. Sync - This option will be synced if user enables sync with all**

**How to prepare profile that DDG toggle in Private NTP is on?**
1. Make clean profile by launching Brave that doesn't have this PR's change
2. Shutdown browser and open `Default/Preferences` file in user dir
3. Put `"use_alternate_private_search_engine":true, "show_alternate_private_search_engine_toggle":true` to `"brave" object` like below.
<img width="825" alt="Screen Shot 2022-06-10 at 12 18 34 PM" src="https://user-images.githubusercontent.com/6786187/172983814-e58bc5ce-c93b-4fb9-953f-b917c290936e.png">

See: https://github.com/brave/brave-browser/issues/18486. Last version with DDG setting was 1.30.89. Removed for new profiles in 1.31.87.
